### PR TITLE
Fixed StagingDeployCash issue body

### LINKED ISSUE #3662: Implement better pagination for our GitHub Actions
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Context
Our GitHub Actions interact with the GitHub API but lack pagination.

## Action Performed:
There's not super clear reproduction steps here, but when [generating the QA/deploy checklist](https://github.com/Expensify/Expensify.cash/blob/266a2a812a702b1500fc26287830aaa5cee4af1b/.github/libs/GithubUtils.js#L196) (always labeled `StagingDeployCash`), we need to cross-check the list of PRs and make sure we _don't_ include PRs created by our CI/CD bot user, `OSBotify`. We do that [here](https://github.com/Expensify/Expensify.cash/blob/266a2a812a702b1500fc26287830aaa5cee4af1b/.github/libs/GithubUtils.js#L203-L218).

However, we only fetch the [latest 100](https://github.com/Expensify/Expensify.cash/blob/266a2a812a702b1500fc26287830aaa5cee4af1b/.github/libs/GithubUtils.js#L207) PRs, while there may be more than 100 PRs merged between deploy cycles (especially including all the automated bot ones). As a result, we include some automated PRs in the deploy checklist, which we do not want to do.

## Expected Result:
We need to fetch all PRs between the oldest merged PR from the list passed in [here](https://github.com/Expensify/Expensify.cash/blob/266a2a812a702b1500fc26287830aaa5cee4af1b/.github/libs/GithubUtils.js#L198) and the newest PR. We can do that by fetching 100 PRs from the API at a time, until the oldest PR we have from the API is older than the oldest PR we were passed in.

## Actual Result:
We don't fetch more than 100 results.

## Workaround:
Ignore or manually delete automated PRs from the deploy checklist.

## Platform:
Where is this issue occurring?

 GitHub only.

view the job on upwork [here](https://www.upwork.com/jobs/~01f64729292b612ace).

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

### PR DESCRIPTION
**Pull request by parasharrajat**
### Details

Added pagination so that all the automated PRs can be filtered while creating the StagingDeployCashBody issue Body.

### Fixed Issues

$ fixes #3662

### Tested On

- [ ] Web
- [ ] Mobile Web
- [ ] Desktop
- [ ] iOS
- [ ] Android

### CODE REVIEW

File: .github/libs/GithubUtils.js [hunk 1]
```diff
@@ -200,13 +200,21 @@ class GithubUtils {
...
-        return this.octokit.pulls.list({
+        return this.octokit.paginate(this.octokit.pulls.list, {
             owner: GITHUB_OWNER,
             repo: EXPENSIFY_CASH_REPO,
             state: 'all',
+            sort: 'created',
+            direction: 'desc',
             per_page: 100,
+        }, ({data}, done) => {
+            const oldestPR = PRList[0];
```
Comment (parasharrajat) [inline .github/libs/GithubUtils.js:211]: I am expecting that PRList is ordered and first element is the oldest. Please tell me If this is incorrect.

**Comment by parasharrajat** [comment]
I think I need help with the Check here. I am not sure exactly why its failing, on local `gh-action-validate` has no errors.

File: .github/libs/GithubUtils.js (hunk 1 above)
Comment (roryabraham) [reply .github/libs/GithubUtils.js:211]: I think that generally with the GitHub API, the first element is the newest. I have this handy bash alias for testing octokit locally:

1. First, create or get an existing personal access token (PAT) for GitHub.
1. Then, put this script in the root of your Expensify.cash folder:

```javascript
const {Octokit} = require('@octokit/rest');

function setupOctokit() {
    return new Octokit({auth: 'INSERT_YOUR_PAT_HERE'});
}

module.exports = setupOctokit;
```

1. (optional) That file should never be added to version control, so add that file to your global gitignore (usually at `~/.gitignore`) so that you don't accidentally commit your PAT.
1. Add this to your shell profile script:

```bash
alias octokit="NODE_PATH=$PATH_TO_E_CASH node -i -e \"let setupOctokit = require( '$PATH_TO_E_CASH/setupOctokit'); let octokit = setupOctokit(); let _ = require('underscore');\""
```

Then you can just type `octokit` in your terminal and it will open an interactive node session with underscore and an authenticated octokit client for you to test read-only commands with.

**Comment by roryabraham** [comment]
> I think I need help with the Check here. I am not sure exactly why its failing, on local `gh-action-validate` has no errors.

**TL;DR** Just run `npm run gh-actions-build` and commit the changes 😄 

But in case you want a more detailed explanation ...

So there are two checks that happen here. `gh-action-validate` uses a json schema to validate our .yaml GitHub actions and workflows to the best of its ability. That script is [here](https://github.com/Expensify/Expensify.cash/blob/c2895982ec799977a233f0cd84925e3246f46710/.github/scripts/validateActionsAndWorkflows.sh).

The other check that happens here requires a bit more context. There is this subtlety of JavaScript GitHub Actions:

> GitHub downloads each action run in a workflow during runtime and executes it as a complete package of code before you can use workflow commands like run to interact with the runner machine. This means you must include any package dependencies required to run the JavaScript code.

So in practice, that means that when creating a JavaScript action, one has two choices:

1. Commit your `node_modules` (or at least, the subset used by your GitHub Actions)
1. Precompile your JavaScript actions with their dependencies into a single Node.js executable using [`ncc`](https://github.com/vercel/ncc).

We didn't want to commit any of our node modules in this repo, so instead we use the second approach. The build script is [here](https://github.com/Expensify/Expensify.cash/blob/658c09479aa95b44d65a0ebd198eae387b72e0fd/.github/scripts/buildActions.sh). You can see the result by looking at all our [JavaScript GitHub Actions](https://github.com/Expensify/Expensify.cash/tree/main/.github/actions), and you'll notice each has:

1. An action metadata `.yml` file
1. A JavaScript source file.
1. A compiled JS file called `index.js`

So whenever changes are made to GH Actions, you need to recompile the GHA executables. In order to help prevent people forgetting to do that, we added another PR check [here](https://github.com/Expensify/Expensify.cash/blob/5627217a238c71c13187fa4abf171f8907c22a9f/.github/workflows/validateGithubActions.yml#L30) that executes [this script](https://github.com/Expensify/Expensify.cash/blob/d4fb326f03fe235600581d58d8e6102d1325abdb/.github/scripts/verifyActions.sh). That script recompiles all the actions on the PR and will complain if there is diff, which is what is happening here:

image

**Comment by parasharrajat** [comment]
Thanks for the detailed Info. but I built actions before pushing. I will try again.

**Comment by roryabraham** [comment]
@parasharrajat Do you have a windows computer?

**Comment by parasharrajat** [comment]
No

**Comment by roryabraham** [comment]
Try running `.github/scripts/verifyActions.sh` locally?

**Comment by roryabraham** [comment]
Okay, next I would try this:

```
rm -rf node_modules
npm i
npm run gh-actions-build
./github/scripts/verifyActions.sh
```

Other than that, you have to look through the diff and try to find out what's going on/why there is a difference. IDK what Linux flavor you have, but that might matter - you could try running the build script in an Ubuntu 20.04 VM?

**Comment by parasharrajat** [comment]
Updated. Thanks.

File: .github/libs/GithubUtils.js [hunk 2]
```diff
@@ -200,13 +200,22 @@ class GithubUtils {
...
+        return this.octokit.paginate(this.octokit.pulls.list, {
             owner: GITHUB_OWNER,
             repo: EXPENSIFY_CASH_REPO,
             state: 'all',
+            sort: 'created',
+            direction: 'desc',
             per_page: 100,
+        }, ({data}, done) => {
+            // PRList is reverse-chronologically ordered
+            const oldestMergedPR = PRList[PRList.length - 1];
```
Comment (roryabraham) [inline .github/libs/GithubUtils.js:212]:
```suggestion
            const oldestMergedPR = _.last(PRList);
```

File: .github/libs/GithubUtils.js [hunk 3]
```diff
@@ -200,13 +200,22 @@ class GithubUtils {
...
+            if (data.find(pr => pr.html_url === oldestMergedPR)) {
```
Comment (roryabraham) [inline .github/libs/GithubUtils.js:213]:
```suggestion
            if (_.find(data, pr => pr.html_url === oldestMergedPR)) {
```

File: .github/libs/GithubUtils.js [hunk 4]
```diff
@@ -200,13 +200,22 @@ class GithubUtils {
...
+                done();
+            }
+            return data;
         })
-            .then(({data}) => {
+            .then((data) => {
```
Comment (roryabraham) [inline .github/libs/GithubUtils.js:218]: Are you sure we don't need this destructuring anymore?

**Review by roryabraham — DISMISSED** [review]
No blockers, just some small comments.

File: .github/libs/GithubUtils.js (hunk 4 above)
Comment (parasharrajat) [reply .github/libs/GithubUtils.js:218]: yes we don't need it

**Comment by parasharrajat** [comment]
Updated. Ready for review.

**Review by roryabraham — APPROVED** [review]
Okay, LGTM 👍
